server/src/main/java/org/opensearch/common/cache/stats/MultiDimensionCacheStats.java

```diff
@@ -0,0 +1,265 @@
+                // Create new node with stats matching the node from the original tree
+                int indexOfLastLevel = dimensionNames.indexOf(levels.get(levels.size() - 1));
+                boolean isLeafNode = depth == indexOfLastLevel; // If this is the last level we aggregate, the new node should be a leaf
+                                                                // node
+                nodeInNewTree = new MDCSDimensionNode(dimensionValue, !isLeafNode, currentInOriginalTree.getStats());
+                parentInNewTree.children.put(dimensionValue, nodeInNewTree);
+            } else {
+                // Otherwise increment existing stats
+                CacheStatsCounterSnapshot newStats = CacheStatsCounterSnapshot.addSnapshots(
+                    nodeInNewTree.getStats(),
+                    currentInOriginalTree.getStats()
+                );
+                nodeInNewTree.setStats(newStats);
+            }
+            // Finally set the parent node to be this node for the next callers of this function
+            parentInNewTree = nodeInNewTree;
+        }
+
+        if (!currentInOriginalTree.children.isEmpty()) {
+            // Not a leaf node
+            for (Map.Entry<String, MDCSDimensionNode> childEntry : currentInOriginalTree.children.entrySet()) {
+                MDCSDimensionNode child = childEntry.getValue();
+                aggregateByLevelsHelper(parentInNewTree, child, levels, depth + 1);
+            }
+        }
+    }
+
+    /**
+     * Filters out levels that aren't in dimensionNames. Unrecognized levels are ignored.
+     */
+    private List<String> filterLevels(List<String> levels) {
+        List<String> filtered = new ArrayList<>();
+        for (String level : levels) {
+            if (dimensionNames.contains(level)) {
+                filtered.add(level);
+            }
+        }
+        if (filtered.isEmpty()) {
+            throw new IllegalArgumentException("Levels cannot have size 0");
+        }
+        return filtered;
+    }
+
+    public CacheStatsCounterSnapshot getStatsForDimensionValues(List<String> dimensionValues) {
+        MDCSDimensionNode current = statsRoot;
+        for (String dimensionValue : dimensionValues) {
+            current = current.children.get(dimensionValue);
+            if (current == null) {
+                return null;
+            }
+        }
+        return current.stats;
+    }
+
+    // A similar class to DimensionNode, which uses an ordered TreeMap and holds immutable CacheStatsCounterSnapshot as its stats.
+    static class MDCSDimensionNode {
+        private final String dimensionValue;
+        final Map<String, MDCSDimensionNode> children; // Map from dimensionValue to the DimensionNode for that dimension value
+
+        // The stats for this node. If a leaf node, corresponds to the stats for this combination of dimensions; if not,
+        // contains the sum of its children's stats.
+        private CacheStatsCounterSnapshot stats;
```

Comment:
I think we should keep stats for inner nodes, because it simplifies aggregation. Having sum-of-children stats means we just have to reconnect nodes to new parents, rather than also adding up values. Plus, we already did the work of computing those stats within StatsHolder.

Reply:
Yeah makes sense. Probably misunderstood this a bit